[JSC] Make isWhiteSpace table-based

Constellation opened the pull request:
#### 3c5cef15cc6ec65d057ef83cb964b099135200c5
<pre>
[JSC] Make isWhiteSpace table-based
<a href="https://bugs.webkit.org/show_bug.cgi?id=288323">https://bugs.webkit.org/show_bug.cgi?id=288323</a>
<a href="https://rdar.apple.com/145439014">rdar://145439014</a>

Reviewed by Yijia Huang.

This patch changes isWhiteSpace table-based implementation since it is
faster than many branches. We also add easy jetstream2-octane-code-load
microbenchmark script.

* JSTests/slowMicrobenchmarks/jetstream2-octane-code-load.js: Added.
(setupCodeLoad):
(tearDownCodeLoad):
(runCodeLoadClosure):
(runCodeLoadJQuery):
(BASE_JS.JQUERY_JS):
(cacheBust):
(runClosure):
(MockElement.this.appendChild):
(MockElement.this.createComment):
(MockElement.this.createDocumentFragment):
(MockElement.this.createElement):
(MockElement.this.getElementById):
(MockElement.this.getElementsByTagName):
(MockElement.this.insertBefore):
(MockElement.this.removeChild):
(MockElement.this.setAttribute):
* Source/JavaScriptCore/parser/Lexer.cpp:
* Source/JavaScriptCore/parser/Lexer.h:
(JSC::Lexer&lt;LChar&gt;::isWhiteSpace):

Canonical link: <a href="https://commits.webkit.org/290932@main">https://commits.webkit.org/290932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aef8ffccc7dd02574ecff5956152ca602c249b80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11036 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96473 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42191 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11414 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70260 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27785 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94505 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82899 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50586 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/486 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41361 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84308 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/491 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98476 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90253 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18666 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79286 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78737 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78490 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23014 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14479 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18662 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23940 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112837 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18373 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32716 "Found 6 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-slow-memory, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-tag-arg.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-tag-arg.js.wasm-slow-memory, wasm.yaml/wasm/stress/top-most-enclosing-stack.js.wasm-eager (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me if bots are happy.

Constellation committed:
[JSC] Make isWhiteSpace table-based
https://bugs.webkit.org/show_bug.cgi?id=288323
rdar://145439014

Reviewed by Yijia Huang.

This patch changes isWhiteSpace table-based implementation since it is
faster than many branches. We also add easy jetstream2-octane-code-load
microbenchmark script.

* JSTests/slowMicrobenchmarks/jetstream2-octane-code-load.js: Added.
(setupCodeLoad):
(tearDownCodeLoad):
(runCodeLoadClosure):
(runCodeLoadJQuery):
(BASE_JS.JQUERY_JS):
(cacheBust):
(runClosure):
(MockElement.this.appendChild):
(MockElement.this.createComment):
(MockElement.this.createDocumentFragment):
(MockElement.this.createElement):
(MockElement.this.getElementById):
(MockElement.this.getElementsByTagName):
(MockElement.this.insertBefore):
(MockElement.this.removeChild):
(MockElement.this.setAttribute):
* Source/JavaScriptCore/parser/Lexer.cpp:
* Source/JavaScriptCore/parser/Lexer.h:
(JSC::Lexer<LChar>::isWhiteSpace):

Canonical link: https://commits.webkit.org/290932@main

kmiller68 commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -42,6 +42,266 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 
 namespace JSC {
 
+// ch == ' ' || ch == '\t' || ch == 0xB || ch == 0xC || ch == 0xA0;
+const uint8_t whiteSpaceTable[256] = {
```
I would have done:

```
static constexpr std::array<uint8_t, 256> makeWhiteSpaceTable() {
    std::array<uint8_t, 256> result { 0 };
    result[' '] = 1;
    result['\t'] = 1;
    result[0xb] = 1;
    result[0xC] = 1;
    result[0xA0] = 1;
    return result;
}

const std::array<uint8_t, 256> whiteSpaceTable = makeWhiteSpaceTable();
```
